x/icqoracle/keeper/icq.go

```diff
@@ -0,0 +1,152 @@
+package keeper
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"time"
+
+	errorsmod "cosmossdk.io/errors"
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	proto "github.com/cosmos/gogoproto/proto"
+
+	"github.com/Stride-Labs/stride/v24/utils"
+	deps "github.com/Stride-Labs/stride/v24/x/icqoracle/deps/types"
+	"github.com/Stride-Labs/stride/v24/x/icqoracle/types"
+	icqtypes "github.com/Stride-Labs/stride/v24/x/interchainquery/types"
+)
+
+const (
+	ICQCallbackID_OsmosisClPool = "osmosisclpool"
+)
+
+// ICQCallbacks wrapper struct for stakeibc keeper
+type ICQCallback func(Keeper, sdk.Context, []byte, icqtypes.Query) error
+
+type ICQCallbacks struct {
+	k         Keeper
+	callbacks map[string]ICQCallback
+}
+
+var _ icqtypes.QueryCallbacks = ICQCallbacks{}
+
+func (k Keeper) ICQCallbackHandler() ICQCallbacks {
+	return ICQCallbacks{k, make(map[string]ICQCallback)}
+}
+
+func (c ICQCallbacks) CallICQCallback(ctx sdk.Context, id string, args []byte, query icqtypes.Query) error {
+	return c.callbacks[id](c.k, ctx, args, query)
+}
+
+func (c ICQCallbacks) HasICQCallback(id string) bool {
+	_, found := c.callbacks[id]
+	return found
+}
+
+func (c ICQCallbacks) AddICQCallback(id string, fn interface{}) icqtypes.QueryCallbacks {
+	c.callbacks[id] = fn.(ICQCallback)
+	return c
+}
+
+func (c ICQCallbacks) RegisterICQCallbacks() icqtypes.QueryCallbacks {
+	return c.
+		AddICQCallback(ICQCallbackID_OsmosisClPool, ICQCallback(OsmosisClPoolCallback))
+}
+
+// Submits an ICQ to get a concentrated liquidity pool from Osmosis' store
+func (k Keeper) SubmitOsmosisClPoolICQ(
+	ctx sdk.Context,
+	tokenPrice types.TokenPrice,
+) error {
+	k.Logger(ctx).Info(utils.LogWithPriceToken(tokenPrice, "Submitting OsmosisClPool ICQ"))
+
+	params := k.GetParams(ctx)
+
+	osmosisPoolId, err := strconv.ParseUint(tokenPrice.OsmosisPoolId, 10, 64)
+	if err != nil {
+		k.Logger(ctx).Error(utils.LogWithPriceToken(tokenPrice, "Error converting osmosis pool id '%s' to uint64, error '%s'", tokenPrice.OsmosisPoolId, err.Error()))
+		return err
+	}
+
+	tokenPriceBz, err := k.cdc.Marshal(&tokenPrice)
+	if err != nil {
+		k.Logger(ctx).Error(utils.LogWithPriceToken(tokenPrice, "Error serializing tokenPrice '%+v' to bytes, error '%s'", tokenPrice, err.Error()))
+		return err
+	}
+
+	queryId := fmt.Sprintf("%s|%s|%s|%d", tokenPrice.BaseDenom, tokenPrice.QuoteDenom, tokenPrice.OsmosisPoolId, ctx.BlockHeight())
+	query := icqtypes.Query{
+		Id:              queryId,
+		ChainId:         params.OsmosisChainId,
+		ConnectionId:    params.OsmosisConnectionId,
+		QueryType:       icqtypes.CONCENTRATEDLIQUIDITY_STORE_QUERY_WITH_PROOF,
+		RequestData:     icqtypes.FormatOsmosisKeyPool(osmosisPoolId),
+		CallbackModule:  types.ModuleName,
+		CallbackId:      ICQCallbackID_OsmosisClPool,
+		CallbackData:    tokenPriceBz,
+		TimeoutDuration: time.Duration(params.IcqTimeoutSec) * time.Second,
+		TimeoutPolicy:   icqtypes.TimeoutPolicy_RETRY_QUERY_REQUEST,
```

Comment:
not sure if we need to have this since I assume we'll just be retrying them right after the timeout

Reply:
be789efa31c628e6faf5eb2d528a6a03201f5a2b